pkg/api/authz.go

```diff
@@ -110,32 +130,78 @@ func (ac *AccessController) isAdmin(username string) bool {
 	return common.Contains(ac.Config.AdminPolicy.Users, username)
 }
 
+func (ac *AccessController) isAnyGroupInAdminPolicy(groups []string, username string) bool {
+	userGroups := ac.getUserGroups(groups, username)
+
+	for _, group := range userGroups {
+		if common.Contains(ac.Config.AdminPolicy.Groups, group) {
+			return true
+		}
+	}
+
+	return false
+}
+
 // getContext builds ac context(allowed to read repos and if user is admin) and returns it.
 func (ac *AccessController) getContext(username string, request *http.Request) context.Context {
-	readGlobPatterns := ac.getGlobPatterns(username, Read)
-	dmcGlobPatterns := ac.getGlobPatterns(username, DetectManifestCollision)
-
-	acCtx := localCtx.AccessControlContext{
-		ReadGlobPatterns: readGlobPatterns,
-		DmcGlobPatterns:  dmcGlobPatterns,
-		Username:         username,
+	acCtx, err := localCtx.GetAccessControlContext(request.Context())
+	if err != nil {
+		return nil
 	}
 
+	readGlobPatterns := ac.getGlobPatterns(username, acCtx.Groups, Read)
+	dmcGlobPatterns := ac.getGlobPatterns(username, acCtx.Groups, DetectManifestCollision)
+
+	acCtx.ReadGlobPatterns = readGlobPatterns
+	acCtx.DmcGlobPatterns = dmcGlobPatterns
+
 	if ac.isAdmin(username) {
 		acCtx.IsAdmin = true
 	} else {
 		acCtx.IsAdmin = false
 	}
 
 	authzCtxKey := localCtx.GetContextKey()
-	ctx := context.WithValue(request.Context(), authzCtxKey, acCtx)
+	ctx := context.WithValue(request.Context(), authzCtxKey, *acCtx)
 
 	return ctx
 }
 
+func (ac *AccessController) getUserGroups(groupNames []string, username string) []string {
```

Comment:
This adds additional groups after the ones already passed as parameter. If we're going to do it this way, we should determine this full list in the authn middleware, and pass it to the context.

Reply:
Now all the groups are passed to the context in authn.go